HTTP Sec-WebSocket-* headers (#35606)
mdn/content: files/en-us/web/api/websockets_api/index.md

- [`MessageEvent`](/en-US/docs/Web/API/MessageEvent)
  - : The event sent by the WebSocket object when a message is received from the server.

> ## Related HTTP headers

Author: FYI Adding headers to the WebSockets API makes them visibly part of the API. Which they are "in effect".

The HTTP headers are used in the [WebSocket handshake](/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake):

- {{HTTPHeader("Sec-WebSocket-Key")}}
  - : An HTTP request header that contains a key from the client.
    This is part of the [WebSocket opening handshake](/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake), and is part of the information used by the server to verify that the client explicitly intends to open a WebSocket.
    It is added automatically by the browser.
- {{HTTPHeader("Sec-WebSocket-Accept")}}
  - : An HTTP {{glossary("response header")}} used in the _WebSocket opening handshake_ to indicate that the server is willing to upgrade to a WebSocket connection.
    The value in the directive is calculated from the value of `Sec-WebSocket-Key` in the corresponding request.

> - {{HTTPHeader("Sec-WebSocket-Version")}}
>   - : An HTTP header that indicates the version of the WebSocket protocol understood by the client in requests.
>     In responses, the presence of the header indicates that the requested version is not supported by the server, and lists the versions that it does support.
> - {{HTTPHeader("Sec-WebSocket-Protocol")}}
>   - : An HTTP header that indicates the sub-protocols supported by the client in requests, and in responses the sub-protocol selected by the server.
> - {{HTTPHeader("Sec-WebSocket-Extensions")}}
>   - : An HTTP header that indicates protocol-level WebSocket extensions supported by the client in requests, and in responses the extension selected by the server.

Author: Matching structure suggested by Brian below.

```suggestion
- {{HTTPHeader("Sec-WebSocket-Version")}}
  - : An HTTP header that in requests indicates the version of the WebSocket protocol understood by the client.
    In responses, the presence of the header indicates that the requested version is not supported by the server, and lists the versions that it does support.
- {{HTTPHeader("Sec-WebSocket-Protocol")}}
  - : An HTTP header that in requests indicates the sub-protocols supported by the client in preferred order.
    In responses, it indicates the the sub-protocol selected by the server from the client's preferences.  
- {{HTTPHeader("Sec-WebSocket-Extensions")}}
  - : An HTTP header that in requests indicates the WebSocket extensions supported by the client in preferred order.
    In responses, it indicates the extension selected by the server from the client's preferences.
```

- [`MessageEvent`](/en-US/docs/Web/API/MessageEvent)
  - : The event sent by the WebSocket object when a message is received from the server.

## Related HTTP headers

The HTTP headers are used in the [WebSocket handshake](/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake):

- {{HTTPHeader("Sec-WebSocket-Key")}}

>   - : An HTTP request header that contains a key from the client.
>     This is part of the [WebSocket opening handshake](/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake), and is part of the information used by the server to verify that the client explicitly intends to open a WebSocket.
>     It is added automatically by the browser.

Reviewer: This might be good to add: it's easy to assume it's got something to do with encrypting a connection, so we might add it here, what do you think:

```suggestion
  - : An HTTP request header that contains a randomly-generated key from the client used in the [WebSocket opening handshake](/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake).
    The key is not an encryption key, it is part of the information used by the server to verify that the client explicitly intends to open a WebSocket.
    It is added automatically by the browser.
```

Author: Good catch. But I think we only need to clarify this because I used the word "key". If I change this to `nonce` then the usage is obvious, and obviously not cryptographic.  Between you and me, this key seems very pointless since it doesn't really tell the server anything much other than that the client can generate a nonce. The client gets back a formatted version, but since it is isn't used as a session key or anything I don't see the value.

```suggestion
  - : An HTTP request header that contains a nonce from the client.
    This is used in the [WebSocket opening handshake](/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake) to verify that the client explicitly intends to open a WebSocket.
    It is added automatically by the browser.
```
